lib/ansible/modules/cloud/vmware/vmware_guest_disk.py

```diff
@@ -0,0 +1,621 @@
+                        self.module.fail_json(msg="Please specify either 'datastore' "
+                                                  "or 'autoselect_datastore' for disk index [%s]" % disk_index)
+
+                    # Check if given value is datastore or datastore cluster
+                    datastore_name = disk['datastore']
+                    datastore_cluster = find_obj(self.content, [vim.StoragePod], datastore_name)
+                    if datastore_cluster:
+                        # If user specified datastore cluster so get recommended datastore
+                        datastore_name = self.get_recommended_datastore(datastore_cluster_obj=datastore_cluster)
+                    # Check if get_recommended_datastore or user specified datastore exists or not
+                    datastore = find_obj(self.content, [vim.Datastore], datastore_name)
+                    if datastore is None:
+                        self.module.fail_json(msg="Failed to find datastore named '%s' "
+                                                  "in given configuration." % disk['datastore'])
+                    current_disk['datastore'] = datastore
+                    current_disk['autoselect_datastore'] = False
+                elif 'autoselect_datastore' in disk:
+                    # Find datastore which fits requirement
+                    datastores = get_all_objs(self.content, [vim.Datastore])
+                    if not datastores:
+                        self.module.fail_json(msg="Failed to gather information about"
+                                                  " available datastores in given datacenter.")
+                    datastore = None
+                    datastore_freespace = 0
+                    for ds in datastores:
+                        if ds.summary.freeSpace > datastore_freespace:
+                            # If datastore field is provided, filter destination datastores
+                            datastore = ds
+                            datastore_freespace = ds.summary.freeSpace
+                    current_disk['datastore'] = datastore
+
+                if 'datastore' not in disk and 'autoselect_datastore' not in disk:
+                    self.module.fail_json(msg="Either 'datastore' or 'autoselect_datastore' is"
+                                              " required parameter while creating disk for "
+                                              "disk index [%s]." % disk_index)
+
+                if [x for x in disk.keys() if x.startswith('size_') or x == 'size']:
+                    # size, size_tb, size_gb, size_mb, size_kb
+                    if 'size' in disk:
+                        size_regex = re.compile(r'(\d+(?:\.\d+)?)([tgmkTGMK][bB])')
+                        disk_size_m = size_regex.match(disk['size'])
+                        if disk_size_m:
+                            expected = disk_size_m.group(1)
+                            unit = disk_size_m.group(2)
```

Comment:
I like the regex approach.

Another modification:
```
if disk_size_m:
   …
```
→
```
if disk_size_m:
   …
else:
   self.module.fail_json(…)
```

(PS: I didn't finish this PR review)

Reply:
Line 446 handles this. But I'm OK adding else block.

Reply:
Thanks for your valuable comments and time. Much appreciated.

Reply:
> Line 446 handles this.

No line 446 handles `TypeError` and `ValueError`. But if disk['size'] don't match the regex, `expected` isn't defined and `if "." in expected:` raise a `NameError` (which you don't catch).

Reply:
Cool.